proc/threads.c

```diff
@@ -1246,29 +1246,60 @@ void proc_threadBroadcastYield(thread_t **queue)
 }
 
 
-int proc_join(time_t timeout)
+int proc_join(int tid, time_t timeout)
 {
-	int err;
-	thread_t *ghost;
-	spinlock_ctx_t sc;
+	int err = EOK, found = 0, id = 0;
 	process_t *process = proc_current()->process;
+	thread_t *ghost = process->ghosts, *firstGhost = process->ghosts, *lastChecked = NULL;
+	spinlock_ctx_t sc;
 
 	hal_spinlockSet(&threads_common.spinlock, &sc);
-	while ((ghost = process->ghosts) == NULL) {
-		err = _proc_threadWait(&process->reaper, timeout, &sc);
 
-		if (err == -EINTR || err == -ETIME)
-			break;
+	if (tid >= 0) {
+		if (firstGhost != NULL) {
+			do {
+				if (ghost->id == tid) {
+					found = 1;
+					break;
+				}
+				ghost = ghost->procnext;
+			} while (ghost != NULL && ghost != firstGhost);
+		}
+		/* wait for thread with `tid` */
+		if (found == 0) {
+			do {
+				err = _proc_threadWait(&process->reaper, timeout, &sc);
+				if (err == -EINTR || err == -ETIME) {
+					break;
+				}
+				firstGhost = process->ghosts;
+				ghost = lastChecked == NULL ? process->ghosts : lastChecked->procnext;
+
+				while (ghost->id != tid && ghost->procnext != firstGhost) {
+					ghost = ghost->procnext;
+				}
+				lastChecked = ghost;
+			} while (ghost != NULL && ghost->id != tid);
+		}
```

Comment:
Hmmm, searching code is implemented 2 times this way. How about alternative algorithm:
do 
 search for id
 if found break
 else threadWait
while (!ETIME & !EINTR)